typescript-requiring-type-checking.js

```diff
@@ -15,23 +15,23 @@ module.exports = {
         /*
             TypeScript rules outside of the recommended configuration
         */
-        // '@typescript-eslint/naming-convention': 'error',
-        // '@typescript-eslint/no-base-to-string': 'error',
-        // '@typescript-eslint/no-confusing-void-expression': 'error',
-        // '@typescript-eslint/no-unnecessary-boolean-literal-compare': 'error',
-        // '@typescript-eslint/no-unnecessary-condition': 'error',
-        // '@typescript-eslint/no-unnecessary-qualifier': 'error',
-        // '@typescript-eslint/no-unnecessary-type-arguments': 'error',
-        // '@typescript-eslint/non-nullable-type-assertion-style': 'error',
-        // '@typescript-eslint/prefer-includes': 'error',
-        // '@typescript-eslint/prefer-nullish-coalescing': 'error',
-        // '@typescript-eslint/prefer-readonly': 'error',
-        // '@typescript-eslint/prefer-readonly-parameter-types': 'error',
-        // '@typescript-eslint/prefer-reduce-type-parameter': 'error',
-        // '@typescript-eslint/prefer-string-starts-ends-with': 'error',
-        // '@typescript-eslint/promise-function-async': 'error',
-        // '@typescript-eslint/require-array-sort-compare': 'error',
-        // '@typescript-eslint/strict-boolean-expressions': 'error',
-        // '@typescript-eslint/switch-exhaustiveness-check': 'error'
+        '@typescript-eslint/naming-convention': 'error',
+        '@typescript-eslint/no-base-to-string': 'error',
+        '@typescript-eslint/no-confusing-void-expression': 'error',
+        '@typescript-eslint/no-unnecessary-boolean-literal-compare': 'error',
+        '@typescript-eslint/no-unnecessary-condition': 'error',
+        '@typescript-eslint/no-unnecessary-qualifier': 'error',
+        '@typescript-eslint/no-unnecessary-type-arguments': 'error',
+        '@typescript-eslint/non-nullable-type-assertion-style': 'error',
+        '@typescript-eslint/prefer-includes': 'error',
+        '@typescript-eslint/prefer-nullish-coalescing': 'error',
+        '@typescript-eslint/prefer-readonly': 'error',
+        '@typescript-eslint/prefer-readonly-parameter-types': 'error',
+        '@typescript-eslint/prefer-reduce-type-parameter': 'error',
+        '@typescript-eslint/prefer-string-starts-ends-with': 'error',
+        '@typescript-eslint/promise-function-async': 'error',
+        '@typescript-eslint/require-array-sort-compare': 'error',
+        '@typescript-eslint/strict-boolean-expressions': 'error',
```

Comment:
And as I demonstrated before, here is where I come in with a [very recent example](https://dev.azure.com/ni/DevCentral/_git/op-web-portal/pullrequest/147444) of where that bit us. ♥

Reply:
I don't think it demonstrates much to produce an even rarer example. That may better demonstrate a lack of testing and review, but I'll leave that argument be because we can't catch every bug. It certainly demonstrates less than all of the cases where we've used unstrict expressions to our benefit, especially checking if a method exists before executing it.

Reply:
@TrevorKarjanis will test how this rule behaves when strict null checks are off. Ideally we'll block numbers from being treated as booleans. We might be able to set that by using allowRuleToRunWithoutStrictNullChecksIKnowWhatIAmDoing and allowNullableNumber but we only want to do that if we're confident it will behave correctly.

Reply:
I think the following is the configuration we want to check (allowing strings but not numbers in boolean expressions):
```js
        '@typescript-eslint/strict-boolean-expressions': ['error', {
            allowString: true,
            allowNumber: false,
            allowNullableString: true,
            allowNullableNumber: false,
            allowRuleToRunWithoutStrictNullChecksIKnowWhatIAmDoing: true,
        }],
```

Reply:
In testing, this rule does not work as expected. I think my suggestion to test was misguided and later corrected by Jesse. The `allowRuleToRunWithoutStrictNullChecksIKnowWhatIAmDoing` option does in fact error to report that strictNullChecks is disabled. In the documentation it states, "If for some reason you cannot turn on strictNullChecks, but still want to use this rule - you can use this option to allow it - but know that the behavior of this rule is undefined with the compiler option turned off." This rule requires strictNullChecks and will report, "Unexpected object value in conditional. The condition is always true." on any object without it, even when combined with an operator, e.g. `if (object && object.length)`.

**Proposal**
Leave this rule disabled, but provide a comment recommending its usage with strictNullChecks. The following is slightly adjusted from Milan's based on default values.
```javascript
/*
    This rule requires the TypeScript strictNullChecks compiler flag. If strictNullChecks is enabled, disallow numbers 
    with the following configuration.
    '@typescript-eslint/strict-boolean-expressions': ['error', {
        allowNumber: false,
        allowNullableBoolean: true,
        allowNullableString: true,
        allowNullableNumber: false
    }]
*/
'@typescript-eslint/strict-boolean-expressions': 'off'
```

Reply:
What is the point then of allowRuleToRunWithoutStrictNullChecksIKnowWhatIAmDoing if it still errors on having strictNullChecks disabled?

Reply:
When false, it ensures the strictNullChecks compiler option is enabled. When true, it allows the rule to be run without strictNullChecks, but its behavior is undefined. I think it ended up in this state, because the author originally expected all users to have strictNullChecks enabled by default if they were using this rule. That seems like a ridiculous assumption given the history of most languages and the compiler's default values, but in the end this rule and some others didn't land in the recommended set for that reason. The authors would prefer to simply [silence the rule](https://github.com/typescript-eslint/typescript-eslint/pull/2345#issuecomment-667384255) if it were ever included in the recommended set.

Reply:
@rajsite this thread is ready to close. @LukasKall hasn't voted on the latest proposal but it's aligned with what we've talked about before so it's likely he's ok with it. If not, we can revisit.